File: app/schemas/recruiter.py
"""Schemas for recruiter dashboard API."""

from datetime import datetime
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class RecruiterReviewState(BaseModel):
    human_review_required: bool = False
    review_status: str = "pending"
    review_notes: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    reviewed_by_user_id: Optional[int] = None


class RecruiterIdentityVerificationMetadata(BaseModel):
    verified: Optional[bool] = None
    confidence_score: Optional[float] = None
    low_identity_confidence: bool = False
    similarity_score: Optional[float] = None
    liveness_mode: Optional[str] = None
    liveness_confidence: Optional[float] = None
    ocr_name: Optional[str] = None
    ocr_document_number: Optional[str] = None
    ocr_confidence: Optional[float] = None
    ocr_name_match: Optional[bool] = None
    message: Optional[str] = None
    warnings: list[str] = Field(default_factory=list)
    evidence_metadata: Optional[dict[str, Any]] = None
    created_at: Optional[datetime] = None


class RecruiterProctorEvent(BaseModel):
    type: str
    severity: str
    time: float
    penalty_percent: float = 0.0
    message: str = ""
    evidence_metadata: Optional[dict[str, Any]] = None


class RecruiterSessionSummary(BaseModel):
    session_id: UUID
    candidate_name: str
    role_title: str
    date: datetime = Field(description="Interview completion time (updated_at)")
    final_score: Optional[float] = None
    recommendation: Optional[str] = None
    status: str
    recording_available: bool = False
    human_review_flag: bool = False
    review_status: str = "pending"
    review_notes: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    integrity_level: Optional[str] = None
    integrity_event_count: int = 0
    low_identity_confidence: bool = False
    invite_token: Optional[str] = None


class RecruiterSessionFilters(BaseModel):
    role_title: Optional[str] = None
    date_from: Optional[datetime] = None
    date_to: Optional[datetime] = None
    invite_token: Optional[str] = None
    score_band: Optional[str] = None
    integrity_level: Optional[str] = None
    review_status: Optional[str] = None


class InviteFunnelMetrics(BaseModel):
    created: int = 0
    opened: int = 0
    registered: int = 0
    verified: int = 0
    started: int = 0
    completed: int = 0


class AssessmentPerformanceMetric(BaseModel):
    token: str
    role_preview: str
    difficulty: str
    question_count: int
    used_count: int
    started_count: int
    completed_count: int
    average_score: Optional[float] = None
    integrity_flag_count: int = 0
    created_at: datetime


class RecruiterAnalyticsResponse(BaseModel):
    generated_at: datetime
    invite_count: int = 0
    completed_session_count: int = 0
    completion_rate_percent: float = 0.0
    average_score: Optional[float] = None
    integrity_flag_rate_percent: float = 0.0
    review_flagged_count: int = 0
    funnel: InviteFunnelMetrics = Field(default_factory=InviteFunnelMetrics)
    score_distribution: dict[str, int] = Field(default_factory=dict)
    integrity_distribution: dict[str, int] = Field(default_factory=dict)
    assessments: list[AssessmentPerformanceMetric] = Field(default_factory=list)


class TranscriptItem(BaseModel):
    index: int
    question: str
    answer: Optional[str] = None
    judgment: Optional[dict[str, Any]] = None
    is_adaptive_follow_up: bool = False
    adaptive_topic: Optional[str] = None
    adaptive_source: Optional[str] = None


class RecruiterSessionDetail(BaseModel):
    session_id: UUID
    candidate_name: str
    role_title: str
    experience_level: str
    status: str
    date: datetime
    created_at: datetime
    duration_minutes: Optional[int] = None
    total_questions: int
    answered_count: int
    final_score: Optional[dict[str, Any]] = None
    original_score: Optional[float] = None
    adjusted_score: Optional[float] = None
    integrity_penalty_percent: float = 0.0
    integrity_level: Optional[str] = None
    integrity_event_count: int = 0
    proctoring_summary: Optional[dict[str, Any]] = None
    low_identity_confidence: bool = False
    identity_similarity_score: Optional[float] = None
    human_review_flag: bool = False
    review_state: RecruiterReviewState
    identity_verification: Optional[RecruiterIdentityVerificationMetadata] = None
    proctor_events: list[RecruiterProctorEvent] = Field(default_factory=list)
    recording_available: bool = False
    recording_filename: Optional[str] = None
    transcript: list[TranscriptItem]
    adaptive_interview: Optional[dict[str, Any]] = None


class HumanReviewUpdateRequest(BaseModel):
    flagged: bool


class RecruiterReviewUpdateRequest(BaseModel):
    review_status: Literal["needs_review", "in_review", "cleared", "escalated", "rejected"]
    review_notes: Optional[str] = None

File: frontend/src/types/recruiter.ts
import type { AnswerJudgment } from "./interview";
import type { FinalScore } from "./interview";
import type { IntegrityViolation } from "./proctor";

export interface RecruiterReviewState {
  human_review_required: boolean;
  review_status: string;
  review_notes: string | null;
  reviewed_at: string | null;
  reviewed_by_user_id: number | null;
}

export interface RecruiterIdentityVerificationMetadata {
  verified?: boolean | null;
  confidence_score?: number | null;
  low_identity_confidence: boolean;
  similarity_score?: number | null;
  liveness_mode?: string | null;
  liveness_confidence?: number | null;
  ocr_name?: string | null;
  ocr_document_number?: string | null;
  ocr_confidence?: number | null;
  ocr_name_match?: boolean | null;
  message?: string | null;
  warnings: string[];
  evidence_metadata?: Record<string, unknown> | null;
  created_at?: string | null;
}

export interface RecruiterProctorEvent extends IntegrityViolation {
  evidence_metadata?: Record<string, unknown> | null;
}

export interface RecruiterSessionSummary {
  session_id: string;
  candidate_name: string;
  role_title: string;
  date: string;
  final_score: number | null;
  recommendation: string | null;
  status: string;
  recording_available: boolean;
  human_review_flag?: boolean;
  review_status: string;
  review_notes?: string | null;
  reviewed_at?: string | null;
  integrity_level?: string | null;
  integrity_event_count: number;
  low_identity_confidence: boolean;
  invite_token?: string | null;
}

export interface RecruiterSessionFilters {
  role_title?: string;
  date_from?: string;
  date_to?: string;
  invite_token?: string;
  score_band?: string;
  integrity_level?: string;
  review_status?: string;
}

export interface InviteFunnelMetrics {
  created: number;
  opened: number;
  registered: number;
  verified: number;
  started: number;
  completed: number;
}

export interface AssessmentPerformanceMetric {
  token: string;
  role_preview: string;
  difficulty: string;
  question_count: number;
  used_count: number;
  started_count: number;
  completed_count: number;
  average_score: number | null;
  integrity_flag_count: number;
  created_at: string;
}

export interface RecruiterAnalyticsResponse {
  generated_at: string;
  invite_count: number;
  completed_session_count: number;
  completion_rate_percent: number;
  average_score: number | null;
  integrity_flag_rate_percent: number;
  review_flagged_count: number;
  funnel: InviteFunnelMetrics;
  score_distribution: Record<string, number>;
  integrity_distribution: Record<string, number>;
  assessments: AssessmentPerformanceMetric[];
}

export interface TranscriptItem {
  index: number;
  question: string;
  answer: string | null;
  judgment: AnswerJudgment | null;
  is_adaptive_follow_up?: boolean;
  adaptive_topic?: string | null;
  adaptive_source?: string | null;
}

export interface AdaptiveInterviewSummary {
  enabled?: boolean;
  prompt_version?: string;
  target_difficulty?: string;
  current_difficulty?: string;
  topics?: string[];
  must_hit_competencies?: string[];
  coverage?: Record<string, number>;
  competency_hits?: Record<string, number>;
  adaptation_count?: number;
  adaptations?: Array<Record<string, unknown>>;
}

export interface RecruiterSessionDetail {
  session_id: string;
  candidate_name: string;
  role_title: string;
  experience_level: string;
  status: string;
  date: string;
  created_at: string;
  duration_minutes: number | null;
  total_questions: number;
  answered_count: number;
  final_score: FinalScore | null;
  original_score: number | null;
  adjusted_score: number | null;
  integrity_penalty_percent: number;
  integrity_level: string | null;
  integrity_event_count: number;
  proctoring_summary: {
    total_violations?: number;
    violations?: IntegrityViolation[];
    score_penalty_percent?: number;
    integrity_level?: string;
  } | null;
  low_identity_confidence?: boolean;
  identity_similarity_score?: number | null;
  human_review_flag?: boolean;
  review_state: RecruiterReviewState;
  identity_verification?: RecruiterIdentityVerificationMetadata | null;
  proctor_events: RecruiterProctorEvent[];
  recording_available: boolean;
  recording_filename: string | null;
  transcript: TranscriptItem[];
  adaptive_interview?: AdaptiveInterviewSummary | null;
}
